airflow/migrations/versions/0027_1_10_0_add_time_zone_awareness.py

```diff
@@ -139,11 +139,11 @@ def upgrade():
             type_=mysql.TIMESTAMP(fsp=6),
         )
     else:
-        # sqlite and mssql datetime are fine as is.  Therefore, not converting
```

Comment:
I am not really sure if we want to remove mssql in all the migrations from the past. 

Maybe yes, Maybe no. I have arguments for both. 

## The argiumets for "removing MSSQL From migrations":

I am leaning towards removing them (so basically what this PR does), but I am a bit torn on that one. On one hand I would love to constraint and limit the "mssql" users to only use Airflow 2.7.3 and never, ever use airflow >= 2.8.0 for anything MSSQL. On the  other hand it feels a little weird. I'd love other's opinion - especially those who worked on the migrations in the past: @ephraimbuddy @dstandish @ashb @kaxil  @eladkal 

## The argument for "leaving MSSQL in migrations": 

It does not really looks like it is necessary to remove them and does not cost us much. It's a bit sunken cost already and we are not going to maintain those migrations anyway. <aybe at some point in time we might want to merge them somehow and get a new "starting point" migration in the future, but I am not sure what is the value in removing mssql from the  old migrations. 

It feels a bit weird to change the past migrations that are already marked as "I run this migration already" by alembic. I know it SHOULD NOT matter when you have Postgres/MySQL, but it feels a bit wrong to modify the migration scripts, especially that it does not hurt us to leave them there (we generally do not touch/maintain migrations too often).

I see two potential prolems with removing the migrations for mssql from those migrations:

* there is a risk we remove something here accidentally by missing it in review - there are a lot of files to be reviewed and lots of conditions, there is a slight risk we will miss it. But with careful review we can mitigate it.

* there is a side-effect on --from-ref/--to-ref, --from-version/--to-version.  Since we are still supporting MSSQL for airlfow < 2.8.0, it would be nice if Airflow 2.8+ still supports migrations for older versions of airflow even for MSSQL.

I think (but I might be wrong) - just leaving the mssql in migration code will just allow it to be run and it SHOULD work in general. So users will still be able - for example - to run `airflow db migrate --to-version 2.7.3` using Airflow 2.8, 2.9 etc. There will be no need to run those migration with `Airflow 2.7.3` strictly.

##  My current thinking

But I am bit torn on that one. It also relates to the discussion we had on https://github.com/apache/airflow/pull/35861 and what  we want to tell our users who run MSSQL now.

So far my points there at least were:

* let's constraint the set of tool and environment as much as we can to reduce variability of environment
* let's provide a tool/script that they can use - but also modify if needed, so that it's not part of "airflow" itself

So if we follow and agree that line of thought - then yes, removing MSSQL from Airflow 2.8.0  and not allowing the MSSQL users to run `--to-version 2.7.3` feels about right. On the othere hand - if we allow for variability, and letting the users to do the migrations in various ways, it's ok to leave it in. 

I wonder what you think?

Reply:
100% on your argument line - and I really had the same thoughts, just did not write it like this.

Mainly I started removing migrations when I cleaned (future) dead code in airflow/migrations/utils.py and then saw migration scripts failing because common functions missing. Migration code is not 100% separated.

As we are shortly before the release and we need more days in discussing/debating about migrations I'd also be OK to split the PR in two. (with a minimum risk that I need to leave some migration dependency dead code behind).
And still if somebody want to play qith MSSQL there is the 2.7.3 release available, from there still you could up-and downgrade. Leaving traces behind if 2.8.0 does not support MSSQL anyway is just some leftover cleanup. And I would not like to keep all MSSQL dependencies in the future (2.9++?) just because in ancient history we supported MSSQL - rather clean it soon.

Reply:
waiting to what others say. I think we do not have to rush it. Spoke to @ephraimbuddy today and RC is planned 11th of December. I think we have quite some time to discuss, decide, iterate and merge.

Reply:
Main risk I see is this PR is quire large, if we merge last-minute, there is no buffer to see if we had a unforeseen side-effect. But I am okay if you see it relaxed :-D

Reply:
Yeah. The problem with size here is that it's "all" or "nothing" in case of migrations (and those are the big ones) . All the rest is "2 minutes" review :). 

So yeah, agree we should - I think - quickly - get to conclusions on how we should do it and start reviewing/merging what we agreed about.

CC: everyone else who migh have an opinion :) . Maybe also worth to send it to devlist  to raise "importance" of it - might be in the response of the vote/consensus we had about MSSQL removal.

Reply:
> On the other hand it feels a little weird

Does it? even before we started experimental support we accepted changes to migration scripts to better support MsSQL. It doesn't cost us anything other than few min of review.
i think we should keep it and also accept PRs improving/fixing issues related to MsSQL.

The removal of MsSQL support from our side just means that we do not test Airflow deployment against MsSQL backend. It doesn't mean that it won't work it just mean that we don't guarantee it.

Reply:
Yeah. It was a "little" weird. But I could go either way.

> The removal of MsSQL support from our side just means that we do not test Airflow deployment against MsSQL backend. It doesn't mean that it won't work it just mean that we don't guarantee it.

I think we need a a bit more than that. 

There are two potential scenarios for our MSSQL current users:

1) Scenario 1 - we do not hard/stop mssql users for migration and it will maybe work and maybe not

If we leave it in "maybe it will work" then our users will continue using airflow until it breaks and willl come to us when it will stop (inevitably) - the whole point is that the "mssql-specific" code in Airflow slows us down, so getting rid of it is a good idea and also it's likely in a few months time, it will catastrophically break MSSQL installation - say somoene already moved to 2.9.0 - still using MsSQL and then it will break in 2.10.0. Now. they have no way to go back (becuase migration scripts will also break for example) and we have a user who has broken installation that they can do nothing about. 

And they will complaint to us, will take our time, will cry for help and workarounds regardless what is the status of MSSQL - and most of all they will loose the trust in Airflow as a good solution for them because it won't work for them for possibly days when they will try to recover somehow.  And this all will be happening 2 years from now when nobody here will even remember what kind of changes they were there implmented and likely many of people who are now in Airflow won't even be around.

And this is actually one of the reasons we want to remove MSSQL - we do not WANT to get issues about MSSQL for 2.8+ 

Is it their fault? Technically yes - they have not read the release notes (I have no stats but there is quite a number of people who don't).

2) Scenario 2 - we forbid MSSQL and remove MSSQL code straight away

We can explicitly forbid migration to 2.8.0 for MSSQL users. that will be much cleaner solution. We will provide them the migration script on 2.7.3 and how to migrate to other database. The users will see it immediatley when they will attempt to migrate - and they will have two choices: a) stay with Airflow 2.7.3 longer b) migrate to other DB.

I think this is far more plausible scenario - both for our users and maintainers (including future ones) of Airflow. The cut is clean, and we know already that (some) people do not read release notes so basically - if we allow MSSQL users to migrate to 2.8.0, then it's all but guaranteed that some of our users will actually get into troubles sooner or later. We are basically accepting the fact that some percentage of our users, rather than stopped from migration - will migrate and eventually get in the troubles.


So I'd personally be for even saying:

```
if mssql => raise Error ("Migrate to Postgres or MySQL").
```